saleor/graphql/core/utils/reordering.py

```diff
@@ -0,0 +1,181 @@
+
+__all__ = ["perform_reordering"]
+
+
+@dataclass(frozen=True)
+class FinalSortOrder:
+    """Describes a final sort order value for a given PK. This is needed to tell django
+    which objects and values to associate and update."""
+
+    pk: int
+    sort_order: int
+
+
+class NoChangeException(Exception):
+    pass
+
+
+class Reordering:
+    def __init__(self, qs: QuerySet, operations: Dict[int, int], field: str):
+        self.qs = qs
+        self.operations = operations
+        self.field = field
+
+        # Will contain the original data, before sorting
+        # This will be useful to look for the sort orders that
+        # actually were changed
+        self.old_sort_map = {}
+
+        # Will contain the list of keys kept
+        # in correct order in accordance to their sort order
+        self.ordered_pks = []
+
+    @cached_property
+    def ordered_node_map(self):
+        ordering_map = OrderedDict(
+            self.qs.select_for_update()
+            .values_list("pk", "sort_order")
+            .order_by(F("sort_order").asc(nulls_last=True), "id")
+        )
+        self.old_sort_map = ordering_map.copy()
+        self.ordered_pks = list(ordering_map.keys())
+
+        previous_sort_order = 0
+
+        # Add sort order to null values
+        for pk, sort_order in ordering_map.items():
+            if sort_order is not None:
+                previous_sort_order = sort_order
+                continue
+
+            previous_sort_order += 1
+            ordering_map[pk] = previous_sort_order
+
+        return ordering_map
+
+    def calculate_new_sort_order(self, pk, move) -> Tuple[int, int, int]:
+        """Returns the proper sort order for the current operation to properly
+        move the node in a given direction with by amount.
+
+        This ensures the new sort order is not biased from gaps between
+        the sort orders."""
+
+        # Skip if we don't have to move it
+        if move == 0:
+            raise NoChangeException(pk, move)
+
+        # Retrieve the position of the node to move
+        node_pos = self.ordered_pks.index(pk)
+
+        # Set the target position from the current position
+        # of the node + the relative position to move from
+        target_pos = node_pos + move
+
+        # Make sure we are not getting out of bounds
+        target_pos = max(0, target_pos)
+        target_pos = min(len(self.ordered_pks) - 1, target_pos)
+
+        # Retrieve the target node and its sort order
+        target_pk = self.ordered_pks[target_pos]
+        target_position = self.ordered_node_map[target_pk]
+
+        # Return the new position
+        return node_pos, target_pos, target_position
+
+    def process_move_operation(self, pk, move):
+        old_sort_order = self.ordered_node_map[pk]
+
+        try:
+            node_pos, target_pos, new_sort_order = self.calculate_new_sort_order(
+                pk, move
+            )
+        except NoChangeException:
```

Comment:
This is fine but I think we could simplify the code if we had something like this:

```
def process_move_operation(self, pk, move):
    if move == 0:
        warnings.warn(...)
        return

    self.calculate_new_sort_order(...)
    # continue with sorting logic
```

This way we're not introducing additional concept such as a custom exception, which is used only in this file.